compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs

```diff
@@ -1286,20 +1285,74 @@ pub fn prohibit_assoc_item_constraint(
         // Check if the type has a generic param with the same name
         // as the assoc type name in the associated item binding.
         let generics = tcx.generics_of(def_id);
-        let matching_param =
-            generics.own_params.iter().find(|p| p.name.as_str() == constraint.ident.as_str());
+        let matching_param = generics.own_params.iter().find(|p| p.name == constraint.ident.name);
 
         // Now emit the appropriate suggestion
         if let Some(matching_param) = matching_param {
-            match (&matching_param.kind, term) {
-                (GenericParamDefKind::Type { .. }, hir::Term::Ty(ty)) => {
-                    suggest_direct_use(&mut err, ty.span);
+            match constraint.kind {
+                hir::AssocItemConstraintKind::Equality { term } => {
+                    match (&matching_param.kind, term) {
+                        (GenericParamDefKind::Type { .. }, hir::Term::Ty(ty)) => {
+                            suggest_direct_use(&mut err, ty.span);
+                        }
+                        (GenericParamDefKind::Const { .. }, hir::Term::Const(c)) => {
+                            let span = tcx.hir().span(c.hir_id);
+                            suggest_direct_use(&mut err, span);
+                        }
+                        _ => suggest_removal(&mut err),
+                    }
                 }
```

Comment:
I wonder if this will read better if you matched `constraint.kind` and `matching_param.kind` together so you could save the indents. For `::Bound` you can just ignore it anyways. So something like

```suggestion
            match (constraint.kind, matching_param.kind) {
                (hir::AssocItemConstraintKind::Equality { term: hir::Term::Ty(ty)  }, GenericParamDefKind::Type { .. }) => {
                     suggest_direct_use(&mut err, ty.span);
```

Reply:
Thanks. The improved code reduces indentation and reads better.